### ISSUE TITLE
Question about the training data for my own TOM network

### ISSUE THREAD

**Issue by YichengZou626** [MIC-DKFZ/TractSeg#225]
Hi, I am trying to train the tractseg model with my own dataset. Right now, I have used MITKFiberDirectionExtraction to get the tom file for each 72 bundles. But since my own data was tck file, the resulted tom file is 145x174x145x9, instead of 145x174x145x3. The code showed that it only accepted the tom file with volume 3, so my first question is whether I can change the number of 3 to 9 in the original code. My second question is that right now I have the folder of 72 tom files for each bundle, should I merge them together to be a single bundle_mask file like previous tract_segmentation network? I use fslmerge, and the final file is 145x174x145x648. I use that file with the peak file to train TOM network, but I got the negative loss during the training. Do you know what caused that?

Here is the log file:
Hyperparameters:
{'BATCH_NORM': False,
 'BATCH_SIZE': 28,
 'BEST_EPOCH': 0,
 'BEST_EPOCH_SELECTION': 'f1',
 'CALC_F1': True,
 'CLASSES': '20',
 'CSD_RESOLUTION': 'LOW',
 'CV_FOLD': 0,
 'DATASET': 'HCP',
 'DATASET_FOLDER': '/nas/longleaf/home/zyc626/TractSeg/tractseg/preproc',
 'DATA_AUGMENTATION': True,
 'DAUG_ALPHA': (90.0, 120.0),
 'DAUG_BLUR_SIGMA': (0, 1),
 'DAUG_ELASTIC_DEFORM': True,
 'DAUG_FLIP_PEAKS': False,
 'DAUG_GAUSSIAN_BLUR': True,
 'DAUG_INFO': '-',
 'DAUG_MIRROR': False,
 'DAUG_NOISE': True,
 'DAUG_NOISE_VARIANCE': (0, 0.05),
 'DAUG_RESAMPLE': False,
 'DAUG_RESAMPLE_LEGACY': False,
 'DAUG_ROTATE': False,
 'DAUG_ROTATE_ANGLE': (-0.2, 0.2),
 'DAUG_SCALE': True,
 'DAUG_SIGMA': (9.0, 11.0),
 'DIM': '2D',
 'DROPOUT_SAMPLING': False,
 'EPOCH_MULTIPLIER': 1,
 'EXPERIMENT_TYPE': 'peak_regression',
 'EXP_MULTI_NAME': '',
 'EXP_NAME': 'my_custom_experiment',
 'EXP_PATH': '/nas/longleaf/home/zyc626/TractSeg/hcp_exp/my_custom_experiment_x5',
 'FEATURES_FILENAME': 'mrtrix_peaks',
 'FLIP_OUTPUT_PEAKS': False,
 'FP16': True,
 'GET_PROBS': True,
 'INFO': '-',
 'INFO_2': 'using AngleLengthLoss, PeakLengthDice',
 'INPUT_DIM': (144, 144),
 'INPUT_RESCALING': False,
 'KEEP_INTERMEDIATE_FILES': False,
 'LABELS_FILENAME': 'bundle_masks',
 'LABELS_FOLDER': 'bundle_masks',
 'LABELS_TYPE': 'float',
 'LEARNING_RATE': 0.001,
 'LOAD_WEIGHTS': False,
 'LOSS_FUNCTION': 'angle_length_loss',
 'LOSS_WEIGHT': 10,
 'LOSS_WEIGHT_LEN': 30,
 'LR_SCHEDULE': True,
 'LR_SCHEDULE_MODE': 'min',
 'LR_SCHEDULE_PATIENCE': 20,
 'METRIC_TYPES': ['loss', 'f1_macro', 'angle_err'],
 'MODEL': 'UNet_Pytorch_DeepSup',
 'MULTI_PARENT_PATH': '/nas/longleaf/home/zyc626/TractSeg/hcp_exp/',
 'NORMALIZE_DATA': True,
 'NORMALIZE_PER_CHANNEL': False,
 'NR_CPUS': -1,
 'NR_OF_CLASSES': 180,
 'NR_OF_GRADIENTS': 9,
 'NR_SLICES': 1,
 'NUM_EPOCHS': 30,
 'ONLY_VAL': False,
 'OPTIMIZER': 'Adamax',
 'OUTPUT_MULTIPLE_FILES': False,
 'PAD_TO_SQUARE': True,
 'PEAK_DICE_LEN_THR': 0.05,
 'PEAK_DICE_THR': [0.95],
 'PREDICT_IMG': False,
 'PREDICT_IMG_OUTPUT': None,
 'PRINT_FREQ': 20,
 'P_SAMP': 1.0,
 'RESET_LAST_LAYER': False,
 'RESOLUTION': '1.25mm',
 'SAVE_WEIGHTS': True,
 'SEGMENT': False,
 'SEG_INPUT': 'Peaks',
 'SLICE_DIRECTION': 'y',
 'SPATIAL_TRANSFORM': 'SpatialTransform',
 'TEST': False,
 'TEST_TIME_DAUG': False,
 'THRESHOLD': 0.5,
 'TRACTSEG_DIR': 'tractseg_output',
 'TRAIN': True,
 'TRAINING_SLICE_DIRECTION': 'y',
 'TYPE': 'single_direction',
 'UNET_NR_FILT': 64,
 'UPSAMPLE_TYPE': 'nearest',
 'USE_DROPOUT': False,
 'USE_VISLOGGER': False,
 'VERBOSE': True,
 'WEIGHTS_PATH': '',
 'WEIGHT_DECAY': 0}
Training...
c0314.ll.unc.edu
Current weight_factor: 10.0
Start looping batches...
train Ep 0, Sp 1, loss 0.001259, t print 97.483s
train Ep 0, Sp 2, loss 0.000272, t print 51.345s
train Ep 0, Sp 3, loss 8.1e-05, t print 50.82s
train Ep 0, Sp 4, loss -0.00035, t print 51.468s
train Ep 0, Sp 5, loss -0.000331, t print 50.729s
train Ep 0, Sp 6, loss -0.000287, t print 50.253s
train Ep 0, Sp 7, loss -0.000193, t print 50.617s
train Ep 0, Sp 8, loss -0.000336, t print 50.59s
train Ep 0, Sp 9, loss -0.000515, t print 51.138s
train Ep 0, Sp 10, loss -0.000463, t print 50.152s
train Ep 0, Sp 11, loss -0.000358, t print 51.76s
train Ep 0, Sp 12, loss -0.000588, t print 50.721s
train Ep 0, Sp 13, loss -0.000654, t print 50.425s
train Ep 0, Sp 14, loss -0.000601, t print 50.393s
train Ep 0, Sp 15, loss -0.000534, t print 50.114s
train Ep 0, Sp 16, loss -0.000518, t print 49.718s
train Ep 0, Sp 17, loss -0.00044, t print 49.625s
train Ep 0, Sp 18, loss -0.000769, t print 49.689s
train Ep 0, Sp 19, loss -0.000718, t print 50.464s
train Ep 0, Sp 20, loss -0.000643, t print 50.543s
train Ep 0, Sp 21, loss -0.000382, t print 50.264s
train Ep 0, Sp 22, loss -0.00069, t print 50.051s
train Ep 0, Sp 23, loss -0.000702, t print 50.057s
train Ep 0, Sp 24, loss -0.000825, t print 50.366s
train Ep 0, Sp 25, loss -0.000536, t print 48.19s
train Ep 0, Sp 26, loss -0.000664, t print 50.365s
train Ep 0, Sp 27, loss -0.000416, t print 50.255s
train Ep 0, Sp 28, loss -0.000921, t print 51.527s
train Ep 0, Sp 29, loss -0.00069, t print 50.58s
train Ep 0, Sp 30, loss -0.000591, t print 51.191s
Start looping batches...

**Comment by wasserth** [comment]
Not sure how you got 9 instead of 3 output channels using MITKFiberDirectionExtraction. Maybe there is a option in MITKFiberDirectionExtraction so specify that you only want one direction (=3 channels). Now you have 3 directions (=9 channels). This can not be used for training. You can only use one direction.
You have to stack all of them into one image. However this would result in 72*3=216 channels, which does not train well. Therefore you have to split it up into four models each with 54 channels. Then you have to train all 4 models separately.

**Comment by YichengZou626** [comment]
I used this command to get the tom files: MitkFiberDirectionExtraction.sh -i /path/tracts/CC.trk -o /path/992774/CC --mask /path/tracts/CC.nii.gz -a 10 -t 0.1 -d 1 -n 3 -f .nii.gz. How to specify that I only want one direction?

**Comment by YichengZou626** [comment]
Also, as you said that I split it up into four models each with 54 channels, and trained 4 models separately. I did not understand after that, how can I use all these trained weights for tracking. If I just used one model's weight, I can only get tracking for 18 bundles. Do you think I should do "Tracking -i peaks.nii.gz" for 4 models separately to get the tracking for all 72 bundles?

**Comment by YichengZou626** [comment]
Got it! So when I use "Tracking -i", what should the command line be like? Does the model automatically include the weights for all 4 models for tracking? I am a little bit confused about how can I just run one tracking for all bundles.
